x509/timestamp_cert_validations.go

```diff
@@ -0,0 +1,164 @@
+			return fmt.Errorf("invalid self-signed certificate. subject: %q. Error: %w", cert.Subject, err)
+		}
+		if err := validateTimestampingLeafCertificate(cert); err != nil {
+			return fmt.Errorf("invalid self-signed certificate. Error: %w", err)
+		}
+		return nil
+	}
+
+	for i, cert := range certChain {
+		if signedTimeError := validateSigningTime(cert, signingTime); signedTimeError != nil {
+			return signedTimeError
+		}
+		if i == len(certChain)-1 {
+			selfSigned, selfSignedError := isSelfSigned(cert)
+			if selfSignedError != nil {
+				return fmt.Errorf("root certificate with subject %q is invalid or not self-signed. Certificate chain must end with a valid self-signed root certificate. Error: %v", cert.Subject, selfSignedError)
+			}
+			if !selfSigned {
+				return fmt.Errorf("root certificate with subject %q is not self-signed. Certificate chain must end with a valid self-signed root certificate", cert.Subject)
+			}
+		} else {
+			// This is to avoid extra/redundant multiple root cert at the end
+			// of certificate-chain
+			selfSigned, selfSignedError := isSelfSigned(cert)
+			// not checking selfSignedError != nil here because we expect
+			// a non-nil err. For a non-root certificate, it shouldn't be
+			// self-signed, hence CheckSignatureFrom would return a non-nil
+			// error.
+			if selfSignedError == nil && selfSigned {
+				if i == 0 {
+					return fmt.Errorf("leaf certificate with subject %q is self-signed. Certificate chain must not contain self-signed leaf certificate", cert.Subject)
+				}
+				return fmt.Errorf("intermediate certificate with subject %q is self-signed. Certificate chain must not contain self-signed intermediate certificate", cert.Subject)
+			}
+			parentCert := certChain[i+1]
+			issuedBy, issuedByError := isIssuedBy(cert, parentCert)
+			if issuedByError != nil {
+				return fmt.Errorf("invalid certificates or certificate with subject %q is not issued by %q. Error: %v", cert.Subject, parentCert.Subject, issuedByError)
+			}
+			if !issuedBy {
+				return fmt.Errorf("certificate with subject %q is not issued by %q", cert.Subject, parentCert.Subject)
+			}
+		}
+
+		if i == 0 {
+			if err := validateTimestampingLeafCertificate(cert); err != nil {
+				return err
+			}
+		} else {
+			if err := validateTimestampingCACertificate(cert, i-1); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func validateTimestampingCACertificate(cert *x509.Certificate, expectedPathLen int) error {
+	if err := validateCABasicConstraints(cert, expectedPathLen); err != nil {
+		return err
+	}
+	return validateTimestampingCAKeyUsage(cert)
+}
+
+func validateTimestampingLeafCertificate(cert *x509.Certificate) error {
+	if err := validateLeafBasicConstraints(cert); err != nil {
+		return err
+	}
+	if err := validateTimestampingKeyUsagePresent(cert); err != nil {
+		return err
+	}
+	if err := validateLeafKeyUsage(cert); err != nil {
+		return err
+	}
```

Comment:
Instead of having the above two methods, consider having a single `validateTimestampingLeafKeyUsage` method. In its implementation, call `validateTimestampingKeyUsagePresent` and `validateLeafKeyUsage`.

Reply:
Same for code signing